File: src/desktop_api/routers/assistant.py
from __future__ import annotations

import logging

from fastapi import APIRouter, Body, Depends, HTTPException, Query

from src.business.services.chat_service import ChatService, DisplayChatMessage
from src.business.services.assistant_failure_service import (
    AssistantRetryConflict,
    AssistantRetryValidation,
    AssistantSessionNotFound,
)
from src.business.agents.tools.clarification_manager import ClarificationValidationError
from src.desktop_api.assistant_runtime import AssistantRuntime
from src.desktop_api.clarifications import (
    pending_clarification_snapshot,
    record_clarification_decision,
)
from src.desktop_api.confirmations import record_confirmation_decision
from src.desktop_api.schemas import (
    AssistantAutoApproveRequest,
    AssistantAutoApproveResponse,
    AssistantClarificationDecisionRequest,
    AssistantClarificationDecisionResponse,
    AssistantClarificationPendingResponse,
    AssistantClarificationSnapshot,
    AssistantConfirmationDecisionRequest,
    AssistantConfirmationDecisionResponse,
    AssistantCreateSessionRequest,
    AssistantCreateSessionResponse,
    AssistantMessage,
    AssistantMessagesResponse,
    AssistantRenameSessionRequest,
    AssistantRetryRequest,
    AssistantRetryResponse,
    AssistantSendMessageRequest,
    AssistantSendMessageResponse,
    AssistantSessionListResponse,
    AssistantSessionSummary,
    AssistantStopRequest,
    AssistantStopResponse,
    AssistantSubagentListResponse,
    AssistantTranscriptResponse,
    ExecutorDetailResponse,
    SegmentBoundaryRequest,
)

router = APIRouter(prefix="/api/assistant", tags=["assistant"])
_runtime: AssistantRuntime | None = None
logger = logging.getLogger(__name__)


def get_chat_service() -> ChatService:
    return ChatService()


def get_assistant_runtime() -> AssistantRuntime:
    global _runtime
    if _runtime is None:
        _runtime = AssistantRuntime()
    return _runtime


def _session_to_dto(item: dict) -> AssistantSessionSummary:
    return AssistantSessionSummary(
        sessionId=item["session_id"],
        title=item["title"],
        preview=item.get("preview") or "",
        status=item.get("status") or "active",
        createdAt=item.get("date"),
        updatedAt=item.get("updated_at"),
        dateLabel=item.get("date_str") or "",
    )


def _message_to_dto(message: DisplayChatMessage) -> AssistantMessage:
    if message.role == "summary":
        role = "summary"
    else:
        role = "assistant" if message.role == "assistant" else "user"
    return AssistantMessage(
        sequence=message.sequence,
        role=role,
        content=message.content,
        createdAt=message.created_at,
        rendering="safe_markdown" if role in ("assistant", "summary") else "plain_text",
        failure=message.failure.to_public_dict() if message.failure is not None else None,
    )


@router.get("")
def assistant_status() -> dict[str, object]:
    return {"status": "ready", "resources": ["sessions", "messages", "confirmations"]}


@router.get("/sessions", response_model=AssistantSessionListResponse)
def list_sessions(
    query: str = "",
    limit: int = Query(200, ge=1, le=500),
    service: ChatService = Depends(get_chat_service),
) -> AssistantSessionListResponse:
    sessions = service.get_sessions_with_preview(limit=limit, query=query)
    return AssistantSessionListResponse(items=[_session_to_dto(item) for item in sessions])


@router.post("/sessions", response_model=AssistantCreateSessionResponse)
def create_session(
    request: AssistantCreateSessionRequest,
    service: ChatService = Depends(get_chat_service),
) -> AssistantCreateSessionResponse:
    try:
        session_id = service.create_session(tool_ids=request.toolIds, title=request.title)
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return AssistantCreateSessionResponse(sessionId=session_id)


@router.patch("/sessions/{session_id}", response_model=AssistantSessionSummary)
def rename_session(
    session_id: str,
    request: AssistantRenameSessionRequest,
    service: ChatService = Depends(get_chat_service),
) -> AssistantSessionSummary:
    try:
        return _session_to_dto(service.rename_session(session_id, request.title))
    except KeyError as exc:
        raise HTTPException(status_code=404, detail="assistant session not found") from exc
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc


@router.delete("/sessions/{session_id}")
def delete_session(
    session_id: str,
    service: ChatService = Depends(get_chat_service),
) -> dict[str, object]:
    if not service.archive_session(session_id):
        raise HTTPException(status_code=404, detail="assistant session not found")
    return {"archived": True, "sessionId": session_id}


@router.get(
    "/sessions/{session_id}/messages",
    response_model=AssistantMessagesResponse,
    response_model_exclude_none=True,
)
def get_messages(
    session_id: str,
    limit: int = Query(10, ge=1, le=100),
    beforeSequence: int | None = Query(default=None, ge=1),
    service: ChatService = Depends(get_chat_service),
) -> AssistantMessagesResponse:
    page = service.get_display_messages(
        session_id,
        limit=limit,
        before_sequence=beforeSequence,
    )
    return AssistantMessagesResponse(
        items=[_message_to_dto(message) for message in page.messages],
        hasMoreBefore=page.has_more_before,
        nextBeforeSequence=page.next_before_sequence,
    )


@router.post("/sessions/{session_id}/messages", response_model=AssistantSendMessageResponse)
def send_message(
    session_id: str,
    request: AssistantSendMessageRequest,
    runtime: AssistantRuntime = Depends(get_assistant_runtime),
) -> AssistantSendMessageResponse:
    try:
        accepted = runtime.dispatch_message(
            session_id,
            request.content,
            continue_subagent=(
                request.continueSubagent.model_dump() if request.continueSubagent else None
            ),
        )
    except ValueError as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return AssistantSendMessageResponse(accepted=accepted, sessionId=session_id)


@router.post(
    "/sessions/{session_id}/retry",
    response_model=AssistantRetryResponse,
)
def retry_message(
    session_id: str,
    request: AssistantRetryRequest,
    runtime: AssistantRuntime = Depends(get_assistant_runtime),
) -> AssistantRetryResponse:
    try:
        accepted = runtime.retry_message(
            session_id,
            request.messageSequence,
            request.content,
        )
    except AssistantSessionNotFound as exc:
        raise HTTPException(status_code=404, detail="assistant session not found") from exc
    except AssistantRetryConflict as exc:
        raise HTTPException(status_code=409, detail=str(exc)) from exc
    except AssistantRetryValidation as exc:
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return AssistantRetryResponse(
        accepted=accepted,
        sessionId=session_id,
        messageSequence=request.messageSequence,
    )


@router.post("/sessions/{session_id}/stop", response_model=AssistantStopResponse)
def stop_session(
    session_id: str,
    request: AssistantStopRequest | None = Body(default=None),
    runtime: AssistantRuntime = Depends(get_assistant_runtime),
) -> AssistantStopResponse:
    """方块停止 = 全停：取消当前回合 + 停掉该会话所有在办用户任务。

    用户心里只有一个"停"——异步派出的执行体靠 session 取消信号穿不透，
    必须经 stop_user_task 的图级取消信号停到。
    """
    accepted = runtime.stop_session_all(
        session_id, run_id=request.runId if request else None
    )
    return AssistantStopResponse(accepted=accepted)


@router.get("/sessions/{session_id}/transcript", response_model=AssistantTranscriptResponse)
def get_transcript(
    session_id: str,
    subagentId: str | None = Query(default=None),
    afterSequence: int | None = Query(default=None),
    beforeSequence: int | None = Query(default=None),
    runtime: AssistantRuntime = Depends(get_assistant_runtime),
) -> AssistantTranscriptResponse:
    """取主助理或指定子任务的过程时间线（含工具，经 009 脱敏；已压缩回合带 compressed 标志）。"""
    try:
        return AssistantTranscriptResponse(
            **runtime.get_transcript(
                session_id,
                subagentId,
                after_sequence=afterSequence,
                before_sequence=beforeSequence,
            )
        )
    except (LookupError, PermissionError) as exc:
        # 归属拒绝（非己出会话的子任务）与缺失一并返回 404「not found」——不确认其存在、
        # 不暴露这是归属问题，符合 FR-030 不泄露他人会话子任务。
        raise HTTPException(status_code=404, detail="subagent transcript not found") from exc
    except Exception as exc:
        logger.error("Assistant transcript failed for session %s: %s", session_id, exc)
        raise HTTPException(status_code=500, detail={"error": "internal_error"}) from exc


@router.get("/sessions/{session_id}/subagents", response_model=AssistantSubagentListResponse)
def list_subagents(
    session_id: str,
    runtime: AssistantRuntime = Depends(get_assistant_runtime),
) -> AssistantSubagentListResponse:
    """取会话子任务权威列表与状态（重连/重开会话兜底）。"""
    try:
        return AssistantSubagentListResponse(**runtime.list_subagents(session_id))
    except Exception as exc:
        logger.error("Assistant subagent list failed for session %s: %s", session_id, exc)
        raise HTTPException(status_code=500, detail={"error": "internal_error"}) from exc


@router.get("/sessions/{session_id}/executor-detail", response_model=ExecutorDetailResponse)
def get_executor_detail(
    session_id: str,
    executorSessionId: str = Query(...),
    runtime: AssistantRuntime = Depends(get_assistant_runtime),
) -> ExecutorDetailResponse:
    """取某执行体会话的聚合详情（⑦ 递归抽屉用）：summary + transcript + children。"""
    try:
        return ExecutorDetailResponse(
            **runtime.get_executor_detail(session_id, executorSessionId)
        )
    except LookupError as exc:
        raise HTTPException(status_code=404, detail="executor not found") from exc
    except Exception as exc:
        logger.error(
            "Executor detail failed for session %s executor %s: %s",
            session_id,
            executorSessionId,
            exc,
        )
        raise HTTPException(status_code=500, detail={"error": "internal_error"}) from exc


@router.post(
    "/confirmations/{request_id}/decision",
    response_model=AssistantConfirmationDecisionResponse,
)
def decide_confirmation(
    request_id: str,
    request: AssistantConfirmationDecisionRequest,
) -> AssistantConfirmationDecisionResponse:
    result = record_confirmation_decision(request_id, request.decision)
    return AssistantConfirmationDecisionResponse(
        requestId=result.request_id,
        decision=result.decision,
        accepted=result.accepted,
    )


@router.get(
    "/sessions/{session_id}/clarifications/pending",
    response_model=AssistantClarificationPendingResponse,
)
def get_pending_clarification(session_id: str) -> AssistantClarificationPendingResponse:
    """返回该会话当前 pending 澄清快照（重连 / 会话打开 / resync 兜底）；无则 clarification=null。"""
    snapshot = pending_clarification_snapshot(session_id)
    return AssistantClarificationPendingResponse(
        clarification=AssistantClarificationSnapshot(**snapshot) if snapshot else None
    )


@router.post(
    "/sessions/{session_id}/clarifications/{request_id}/decision",
    response_model=AssistantClarificationDecisionResponse,
)
def decide_clarification(
    session_id: str,
    request_id: str,
    request: AssistantClarificationDecisionRequest,
) -> AssistantClarificationDecisionResponse:
    try:
        result = record_clarification_decision(
            session_id,
            request_id,
            request.decision,
            [a.model_dump() for a in request.answers],
        )
    except LookupError as exc:
        # 归属错配/缺失一并 404，不泄漏存在性
        raise HTTPException(status_code=404, detail="clarification not found") from exc
    except ClarificationValidationError as exc:
        # 答案校验失败：422，不结算，前端保留卡片可重试
        raise HTTPException(status_code=422, detail=str(exc)) from exc
    return AssistantClarificationDecisionResponse(
        requestId=result["requestId"],
        status=result["status"],
        accepted=result["accepted"],
    )


@router.post(
    "/confirmations/auto-approve",
    response_model=AssistantAutoApproveResponse,
)
def set_auto_approve(
    request: AssistantAutoApproveRequest,
    runtime: AssistantRuntime = Depends(get_assistant_runtime),
) -> AssistantAutoApproveResponse:
    return AssistantAutoApproveResponse(enabled=runtime.set_auto_approve(request.enabled))


@router.post("/sessions/{session_id}/segment-idle")
def trigger_segment_idle(session_id: str):
    """前端空闲计时器触发后调用，封存当前 Segment"""
    from src.business.brain.segment_service import SegmentService

    try:
        segment_id = SegmentService().handle_idle_and_cleanup(session_id)
    except Exception as exc:
        logger.error("Segment idle trigger failed for session %s: %s", session_id, exc)
        raise HTTPException(status_code=500, detail={"error": "internal_error"}) from exc
    if segment_id is None:
        return {"segment_id": None, "status": None}
    return {"segment_id": segment_id, "status": "pending"}


@router.post("/segment-boundary")
def trigger_segment_boundary(body: SegmentBoundaryRequest):
    """通用 Segment 边界触发端点（window_close / new_session / token_limit）"""
    from src.business.brain.segment_service import SegmentService

    try:
        segment_id = SegmentService().seal_and_cleanup(body.session_id, boundary_reason=body.reason)
    except Exception as exc:
        logger.error("Segment boundary trigger failed for session %s: %s", body.session_id, exc)
        raise HTTPException(status_code=500, detail={"error": "internal_error"}) from exc
    return (
        {"segment_id": segment_id, "status": "pending"}
        if segment_id
        else {"segment_id": None, "status": None}
    )

File: frontend/src/api/assistant.ts
import { requestJson } from "./client";

export interface AssistantSession {
  sessionId: string;
  title: string;
  preview: string;
  status: string;
  createdAt: string | null;
  updatedAt: string | null;
  dateLabel: string;
}

export interface AssistantMessage {
  sequence: number;
  role: "user" | "assistant" | "summary";
  content: string;
  createdAt: string | null;
  rendering: "plain_text" | "safe_markdown";
}

export interface AssistantMessagesResponse {
  items: AssistantMessage[];
  hasMoreBefore: boolean;
  nextBeforeSequence: number | null;
}

export interface AssistantConfirmation {
  requestId: string;
  sessionId?: string;
  actionType: "write_file" | "edit_file" | "exec" | "skill.edit_protected" | "skill.soft_delete" | "unknown";
  sanitizedSummary: string;
  status: "queued" | "active" | "approved" | "denied" | "timed_out" | "cancelled";
  expiresAt?: string | null;
  affectedSkillId?: string;
  affectedEquipmentCount?: number;
  affectedSpecialistNames?: string[];
}

export async function listAssistantSessions(query = "", limit = 200): Promise<AssistantSession[]> {
  const params = new URLSearchParams({ limit: String(limit) });
  if (query.trim()) {
    params.set("query", query.trim());
  }
  const response = await requestJson<{ items: AssistantSession[] }>(`/api/assistant/sessions?${params}`);
  return response.items ?? [];
}

export async function createAssistantSession(input: {
  title?: string;
  toolIds?: string[];
} = {}): Promise<string> {
  const response = await requestJson<{ sessionId: string }>("/api/assistant/sessions", {
    method: "POST",
    body: JSON.stringify(input),
  });
  return response.sessionId;
}

export async function renameAssistantSession(sessionId: string, title: string): Promise<AssistantSession> {
  return requestJson<AssistantSession>(`/api/assistant/sessions/${encodeURIComponent(sessionId)}`, {
    method: "PATCH",
    body: JSON.stringify({ title }),
  });
}

export async function deleteAssistantSession(sessionId: string): Promise<void> {
  await requestJson<{ archived: boolean }>(`/api/assistant/sessions/${encodeURIComponent(sessionId)}`, {
    method: "DELETE",
  });
}

export function listAssistantMessages(
  sessionId: string,
  options: { limit?: number; beforeSequence?: number } = {},
): Promise<AssistantMessagesResponse> {
  const params = new URLSearchParams({ limit: String(options.limit ?? 10) });
  if (options.beforeSequence !== undefined) {
    params.set("beforeSequence", String(options.beforeSequence));
  }
  return requestJson<AssistantMessagesResponse>(
    `/api/assistant/sessions/${encodeURIComponent(sessionId)}/messages?${params}`,
  );
}

export function sendAssistantMessage(
  sessionId: string,
  content: string,
  options: {
    continueSubagent?: {
      subagentId: string;
      supplemental?: string;
    };
  } = {},
): Promise<{ accepted: boolean; sessionId: string }> {
  return requestJson<{ accepted: boolean; sessionId: string }>(
    `/api/assistant/sessions/${encodeURIComponent(sessionId)}/messages`,
    {
      method: "POST",
      body: JSON.stringify({
        content,
        ...(options.continueSubagent ? { continueSubagent: options.continueSubagent } : {}),
      }),
    },
  );
}

export function stopAssistantRun(sessionId: string, runId?: string): Promise<{ accepted: boolean }> {
  return requestJson<{ accepted: boolean }>(
    `/api/assistant/sessions/${encodeURIComponent(sessionId)}/stop`,
    {
      method: "POST",
      ...(runId ? { body: JSON.stringify({ runId }) } : {}),
    },
  );
}

export interface AssistantActivityStep {
  kind: "reasoning" | "tool_call" | "tool_result";
  toolName?: string | null;
  text: string;
  seq: number;
}

export interface AssistantTranscript {
  steps: AssistantActivityStep[];
  compressed: boolean;
}

export function getSubagentTranscript(
  sessionId: string,
  subagentId?: string,
  options: { afterSequence?: number; beforeSequence?: number } = {},
): Promise<AssistantTranscript> {
  const params = new URLSearchParams();
  if (subagentId) {
    params.set("subagentId", subagentId);
  }
  if (options.afterSequence !== undefined) {
    params.set("afterSequence", String(options.afterSequence));
  }
  if (options.beforeSequence !== undefined) {
    params.set("beforeSequence", String(options.beforeSequence));
  }
  const query = params.toString();
  return requestJson<AssistantTranscript>(
    `/api/assistant/sessions/${encodeURIComponent(sessionId)}/transcript${query ? `?${query}` : ""}`,
  );
}

export type SubagentStatus = "running" | "done" | "suspended" | "failed";

export interface AssistantSubagentSummary {
  subagentId: string;
  label: string;
  task: string;
  status: SubagentStatus;
  lastOutput?: string | null;
  turnStartSequence?: number | null;
}

export async function listSubagents(sessionId: string): Promise<AssistantSubagentSummary[]> {
  const response = await requestJson<{ items: AssistantSubagentSummary[] }>(
    `/api/assistant/sessions/${encodeURIComponent(sessionId)}/subagents`,
  );
  return response.items ?? [];
}

export function decideAssistantConfirmation(
  requestId: string,
  decision: "approve" | "deny",
): Promise<{ requestId: string; decision: "approve" | "deny"; accepted: boolean }> {
  return requestJson<{ requestId: string; decision: "approve" | "deny"; accepted: boolean }>(
    `/api/assistant/confirmations/${encodeURIComponent(requestId)}/decision`,
    {
      method: "POST",
      body: JSON.stringify({ decision }),
    },
  );
}

export function triggerAssistantSegmentIdle(
  sessionId: string,
): Promise<{ segment_id: string | null; status: string | null }> {
  return requestJson<{ segment_id: string | null; status: string | null }>(
    `/api/assistant/sessions/${encodeURIComponent(sessionId)}/segment-idle`,
    { method: "POST" },
  );
}

export function triggerAssistantSegmentBoundary(
  sessionId: string,
  reason: "window_close" | "new_session" | "token_limit",
): Promise<{ segment_id: string | null; status: string | null }> {
  return requestJson<{ segment_id: string | null; status: string | null }>("/api/assistant/segment-boundary", {
    method: "POST",
    body: JSON.stringify({ session_id: sessionId, reason }),
  });
}

export async function setAssistantAutoApprove(enabled: boolean): Promise<boolean> {
  const response = await requestJson<{ enabled: boolean }>("/api/assistant/confirmations/auto-approve", {
    method: "POST",
    body: JSON.stringify({ enabled }),
  });
  return response.enabled;
}
